uptane/clients/secondary.py

```diff
@@ -544,20 +562,221 @@ def get_validated_target_info(self, target_filepath):
+
+      tuf.ExpiredMetadataError
+        if the Targets metadata is expired
+
+      tuf.ReplayedMetadataError
+        if the Targets metadata has a lower version number than
+        the last Targets metadata this client deemed valid (rollback)
+
+    <Side-Effects>
+      May update this client's metadata (Director Targets); see <Purpose>
+      May update self.validated_targets_for_this_ecu; see <Purpose>
+
+    """
+    tuf.formats.RELPATH_SCHEMA.check_match(director_targets_metadata_fname)
+    # Checks if the secondary holds the director's public key
+    if self.director_public_key is None:
+      raise uptane.Error("Director public key not found for partial"
+          " verification of secondary.")
+    validated_targets_for_this_ecu = []
+    target_metadata = {}
+    if not os.path.exists(director_targets_metadata_fname):
+      raise uptane.Error('Indicated metadata archive does not exist. '
+          'Filename: ' + repr(director_targets_metadata_fname))
+
+    metadata_file_object = tuf.util.load_file(director_targets_metadata_fname)
+
+    data = metadata_file_object['signed']
+
+
+    # Check to see if the metadata is expired by comparing it against the
+    # last validated time provided by the timeserver. (This may be old, and
+    # we take it as a minimum time.)
+    minimum_time = tuf.formats.datetime_to_unix_timestamp(iso8601.parse_date(
+        self.all_valid_timeserver_times[-1]))
+
+    expiration_time = tuf.formats.datetime_to_unix_timestamp(iso8601.parse_date(
+        data['expires']))
+
+    if expiration_time < minimum_time:
+      raise tuf.ExpiredMetadataError('Expired metadata provided to partial '
+          'verification Secondary; last valid timeserver time: ' +
+          self.all_valid_timeserver_times[-1] + '; expiration date on '
+          'metadata: ' + data['expires'])
+
+
+    # Check to see if the metadata has a lower version number than expected.
+    if data['version'] < self.last_valid_targets_metadata_version_number:
+      raise tuf.ReplayedMetadataError('Provided metadata has lower version '
+          'number than the metadata this partial verification Secondary has '
+          'previously validated.')
+
+
+    # Make sure the data is in the exact format that it is expected to have
+    # been signed over in order to validate the signature over it.
+    # - In the case of JSON, that means TUF's canonical JSON encoding.
+    # - In the case of ASN.1/DER, that means a hash taken over the ASN.1/DER
+    #   data (so it must be converted back, as the data was converted into a
+    #   dictionary for ease of use).
+    if director_targets_metadata_fname.endswith('json'):
+      data_signed = tuf.formats.encode_canonical(data).encode('utf-8')
+
+    elif director_targets_metadata_fname.endswith('der'):
+      data_signed = tuf_asn1_codec.convert_signed_metadata_to_der(
+          {'signed': data, 'signatures': []}, only_signed=True)
+      data_signed = hashlib.sha256(data_signed).digest()
```

Comment:
I think code in the reference implementation elsewhere always depends on tuf.conf.METADATA_FORMAT. The Secondary should be using that to determine what kind of file it's dealing with, not basing it on the file ending, I think.